xgi/classes/dihypergraph.py

```diff
@@ -0,0 +1,610 @@
+    >>> H.nodes
+    NodeView((1, 2, 3, 4, 5, 6, 7, 8))
+    >>> H.edges
+    EdgeView((0, 1, 2, 3))
+
+    """
+    _node_dict_factory = IDDict
+    _node_attr_dict_factory = IDDict
+    _hyperedge_dict_factory = IDDict
+    _hyperedge_attr_dict_factory = IDDict
+    _hypergraph_attr_dict_factory = dict
+
+    def __init__(self, **attr):
+        self._edge_uid = count()
+        self._hypergraph = self._hypergraph_attr_dict_factory()
+        self._node_in = self._node_dict_factory()
+        self._node_out = self._node_dict_factory()
+        self._node_attr = self._node_attr_dict_factory()
+        self._edge_in = self._hyperedge_dict_factory()
+        self._edge_out = self._hyperedge_dict_factory()
+        self._edge_attr = self._hyperedge_attr_dict_factory()
+
+        self._hypergraph.update(attr)  # must be after convert
+
+    @property
+    def num_nodes(self):
+        """The number of nodes in the hypergraph.
+
+        Returns
+        -------
+        int
+            The number of nodes in the hypergraph.
+
+        See Also
+        --------
+        num_edges : returns the number of edges in the hypergraph
+
+        Examples
+        --------
+        >>> import xgi
+        >>> hyperedge_list = [[1, 2], [2, 3, 4]]
+        >>> H = xgi.Hypergraph(hyperedge_list)
+        >>> H.num_nodes
+        4
+
+        """
+        return len(self._node_in)
+
+    @property
+    def num_edges(self):
+        """The number of edges in the hypergraph.
+
+        Returns
+        -------
+        int
+            The number of edges in the hypergraph.
+
+        See Also
+        --------
+        num_nodes : returns the number of nodes in the hypergraph
+
+        Examples
+        --------
+        >>> import xgi
+        >>> hyperedge_list = [[1, 2], [2, 3, 4]]
+        >>> H = xgi.Hypergraph(hyperedge_list)
+        >>> H.num_edges
+        2
+
+        """
+        return len(self._edge_in)
+
+    def add_node(self, node, **attr):
+        """Add one node with optional attributes.
+
+        Parameters
+        ----------
+        node : node
+            A node can be any hashable Python object except None.
+        attr : keyword arguments, optional
+            Set or change node attributes using key=value.
+
+        See Also
+        --------
+        add_nodes_from
+
+        Notes
+        -----
+        If node is already in the hypergraph, its attributes are still updated.
+
+        """
+        if node not in self._node_in:
+            self._node_in[node] = set()
+            self._node_out[node] = set()
+            self._node_attr[node] = self._node_attr_dict_factory()
+        self._node_attr[node].update(attr)
+
+    def add_nodes_from(self, nodes_for_adding, **attr):
+        """Add multiple nodes with optional attributes.
+
+        Parameters
+        ----------
+        nodes_for_adding : iterable
+            An iterable of nodes (list, dict, set, etc.).
```

Comment:
Personally, I don't like that we have to specify what an iterable is. The parenthesis also says that a dict is possible, but it's not clear what the keys and values o such a dict are. We can leave this for a different issue tho.

Reply:
I agree. I will open as an issue.